### ISSUE TITLE
Can rootless and rooted both docker use GPU without changing no-cgroups?

### ISSUE THREAD

**Issue by santosh** [NVIDIA/nvidia-container-toolkit#85]
### 1. Issue or feature description

Both rootless and rooted docker can't seem to access GPU without setting a config in `/etc/nvidia-container-runtime/config.toml`. 

The setting needs to be modified back and forth for rootless and rooted docker to access GPU respectively.

### 2. Steps to reproduce the issue

Assuming that you already have the rootful version of docker.

1. Install nvidia-continer-runtime ([docs](https://nvidia.github.io/nvidia-container-runtime/)).
2. Run `docker run --rm --gpus all nvidia/cuda:12.0.0-base-ubuntu20.04 nvidia-smi`, and you should see the output.
3. Install rootless docker: `dockerd-rootless-setuptool.sh install`. Start the user level service.
4. Run this command to execute nvidia-smi inside a rooless container: `DOCKER_HOST=unix://$XDG_RUNTIME_DIR/docker.sock docker run --rm --gpus all nvidia/cuda:12.0.0-base-ubuntu20.04 nvidia-smi`. It should fail with:

```
docker: Error response from daemon: failed to create task for container: failed to create shim task: OCI runtime create failed: runc create failed: unable to start container process: error during container init: error running hook #0: error running hook: exit status 1, stdout: , stderr: Auto-detected mode as 'legacy'
nvidia-container-cli: mount error: failed to add device rules: unable to find any existing device filters attached to the cgroup: bpf_prog_query(BPF_CGROUP_DEVICE) failed: operation not permitted: unknown.
```

5. This error can easily be fixed by setting `no-cgroups = true` in `/etc/nvidia-container-runtime/config.toml`. This will let the nvidia-smi run normally. As suggested in https://github.com/NVIDIA/nvidia-docker/issues/1155.
6. With change made in above step. Run this command to execute nvidia-smi in rooted docker: `DOCKER_HOST=unix:///var/run/docker.sock docker run --rm --gpus all nvidia/cuda:12.0.0-base-ubuntu20.04 nvidia-smi`. It should fail with:

```
Failed to initialize NVML: Unknown Error
```
7. If you revert the settings, you'll be back to failed operation on rootless and successful on rooted.

This indicates that both the rootless docker and normal docker can't access GPU at the same time.

Is there any solution for that?

**Comment by santosh** [comment]
We don't intend to switch to podman right now.

**Comment by elezar** [comment]
@santosh the upcoming Docker 25 release will include CDI support if the `experimental` flag is enabled in the Daemon.

The documentation is under review and will be published alongside our NVIDIA Container Toolkit 1.14.0 release.

**Comment by davidshen84** [comment]
I ran a rooted docker on my WSL 2 environment and got this error recently.

```
Failed to initialize NVML: GPU access blocked by the operating system                                                                                                                                                                                       Failed to properly shut down NVML: GPU access blocked by the operating system
```

I am not sure if it is because I added this to my WSL kernel command:

```
kernelCommandLine = cgroup_no_v1=all systemd.unified_cgroup_hierarchy=1
```

I have set `no-cgroups=false` in the `/etc/nvidia-container-runtime/config.toml` file to let my docker access my GPU in WSL again.

So, what does this `no-cgroups` mean in the `config.toml` file? I could not find any documentation about it.


Thanks.

**Comment by vsoch** [comment]
@elezar is CDI supported for Docker 28.0.0 now? I am having this specific issue (where I can't use no-cgroups) and would like to test CDI - my setup is using rootless docker and docker compose.

```console
Client: Docker Engine - Community
 Version:    28.0.0
 Context:    default
 Debug Mode: false
 Plugins:
  buildx: Docker Buildx (Docker Inc.)
    Version:  v0.21.0
    Path:     /usr/libexec/docker/cli-plugins/docker-buildx
  compose: Docker Compose (Docker Inc.)
    Version:  v2.33.0
    Path:     /usr/libexec/docker/cli-plugins/docker-compose

Server:
 Containers: 1
  Running: 0
  Paused: 0
  Stopped: 1
 Images: 2
 Server Version: 28.0.0
 Storage Driver: vfs
 Logging Driver: json-file
 Cgroup Driver: systemd
 Cgroup Version: 2
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local splunk syslog
 Swarm: inactive
 Runtimes: io.containerd.runc.v2 runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: bcc810d6b9066471b0b6fa75f557a15a1cbf31bb
 runc version: v1.2.4-0-g6c52b3f
 init version: de40ad0
 Security Options:
  seccomp
   Profile: builtin
  rootless
  cgroupns
 Kernel Version: 5.10.0-33-cloud-amd64
 Operating System: Debian GNU/Linux 11 (bullseye)
 OSType: linux
 Architecture: x86_64
 CPUs: 16
 Total Memory: 58.87GiB
 Name: flux-001
 ID: 4677b6dc-8bb1-46bc-ae97-63bad52dc8e6
 Docker Root Dir: /home/jupyter/.local/share/docker
 Debug Mode: false
 Experimental: false
 Insecure Registries:
  ::1/128
  127.0.0.0/8
 Live Restore Enabled: false
```

**Comment by ntsarb** [comment]
I was in the process of deploying NVIDIA's libraries on WSL2 Ubuntu 22.04 (Windows 11 Pro).

After installing the NVIDIA Container from here:
https://docs.nvidia.com/datacenter/cloud-native/container-toolkit/latest/install-guide.html#installing-the-nvidia-container-toolkit

I was instructed to test it:
(Source: https://docs.nvidia.com/datacenter/cloud-native/container-toolkit/latest/sample-workload.html)

sudo docker run --rm --runtime=nvidia --gpus all ubuntu nvidia-smi

_Failed to initialize NVML: GPU access blocked by the operating system
Failed to properly shut down NVML: GPU access blocked by the operating system_

I found the workaround here:
https://stackoverflow.com/questions/72932940/failed-to-initialize-nvml-unknown-error-in-docker-after-few-hours

I checked the contents of the file /etc/nvidia-container-runtime/config.toml and found it to have:
no-cgroups = true

Changing the value to false resolved the issue.

**Comment by vsoch** [comment]
Sorry I didn't report back! I got this working end of February, wrote up details here.

https://vsoch.github.io/2025/rootless-usernetes-gpu/

It's not just a matter of setting that value to false because we have two layers - the usernetes "node" (rootless container) and then the containerd running inside.

**Comment by elezar** [comment]
As a note. The latest version of the NVIDIA Container Toolkit uses CDI which SHOULD provide better support for rootless and rootful containers. (The no-cgroups setting is not used when the `nvidia` runtime is used).

I am closing this issue. If any problems persist using the latest version of the NVIDIA Container Toolkit, please open a new issue.

**Comment by acottre** [comment]
Thanks, this has fixed the issue for me. Should the third point in the docs at https://docs.nvidia.com/datacenter/cloud-native/container-toolkit/latest/install-guide.html#rootless-mode be removed?

>Configure /etc/nvidia-container-runtime/config.toml by using the sudo nvidia-ctk command:

`sudo nvidia-ctk config --set nvidia-container-cli.no-cgroups --in-place`
